src/zenml/zen_server/zen_server_api.py

```diff
@@ -159,6 +164,34 @@ async def set_secure_headers(request: Request, call_next: Any) -> Any:
     return response
 
 
+@app.middleware("http")
+async def track_last_user_activity(request: Request, call_next: Any) -> Any:
+    """A middleware to track last user activity.
+
+    This middleware checks if the incoming request is a user request and
+    updates the last activity timestamp if it is.
+
+    Args:
+        request: The incoming request object.
+        call_next: A function that will receive the request as a parameter and
+            pass it to the corresponding path operation.
+
+    Returns:
+        The response to the request.
+    """
+    global last_user_activity
```

Comment:
Not a big fan of this approach. Can we use DB for that? The overhead should be quite minimal, just ensure you write in async mode.

Reply:
@avishniakov why are you not a big fan? Why do i need to do a DB write? whats the argument against this in memory?

Reply:
As raised by @stefannica, memory is not syncing across nodes and is tending to errors on restarts. Both arguments make this variable a candidate for permanent storage with sync across nodes.

For your case specifically, imagine a server running 1 node and outing every 1 hour due to a memory leak or whatever. For you this server will be forever active, while it should be shut down after 12 hours. Is this fine with you? A DB is a simple improvement which can solve it in one shot and with minimal overhead, IMO.

Reply:
alright agreed